src/math/quaternion.lisp

```diff
@@ -9,6 +9,103 @@
 (defun make-quaternion (&key (x 0.0f0) (y 0.0f0) (z 0.0f0) (w 1.0f0))
   (make-instance 'quaternion :x x :y y :z z :w w))
 
+(defmacro make-quaternion-f2 (f2)
+  `(make-quaternion :w 0 :x (float2-x ,f2) :y (float2-y ,f2) :z 0))
+
+(defmacro make-quaternion-f3 (f3)
+  `(make-quaternion :w 0 :x (float3-x ,f3) :y (float3-y ,f3) :z (float3-z ,f3)))
+
+(defun quaternion-zero () (make-quaternion :x 0 :y 0 :z 0 :w 0))
+(defun quaternion-one () (make-quaternion :x 1 :y 1 :z 1 :w 1))
+
+(defmacro quaternion-v (q)
+  `(make-float3 (quaternion-x ,q)
+                (quaternion-y ,q)
+                (quaternion-z ,q)))
+
+(defmacro quaternion-a (q)
+  `(quaternion-w ,q))
+
+(defmethod print-object ((q quaternion) out)
+  (let* ((w (quaternion-w q))
+         (x (quaternion-x q))
+         (y (quaternion-y q))
+         (z (quaternion-z q))
+         (xsgn (get-sign x))
+         (ysgn (get-sign y))
+         (zsgn (get-sign z)))
+      (format out "~a ~a ~ai ~a ~aj ~a ~ak" w xsgn (abs x) ysgn (abs y) zsgn (abs z))))
+
+(defmacro define-quaternion-op (name op)
+  `(defmethod ,name ((q1 quaternion) (q2 quaternion))
+    (make-quaternion
+      :x (funcall ,op (quaternion-x q1) (quaternion-x q2))
+      :y (funcall ,op (quaternion-y q1) (quaternion-y q2))
+      :z (funcall ,op (quaternion-z q1) (quaternion-z q2))
+      :w (funcall ,op (quaternion-w q1) (quaternion-w q2)))))
+
+(defmacro define-quaternion-num-op (name op)
+  `(defmethod ,name ((q quaternion) (num real))
+    (make-quaternion
+      :x (funcall ,op (quaternion-x q) num)
+      :y (funcall ,op (quaternion-y q) num)
+      :z (funcall ,op (quaternion-z q) num)
+      :w (funcall ,op (quaternion-w q) num))))
+
+(defmacro define-quaternion-num-op-revord (name op)
+  `(defmethod ,name ((num real) (q quaternion))
+    (make-quaternion
+     :x (funcall ,op num (quaternion-x q))
+     :y (funcall ,op num (quaternion-y q))
+     :z (funcall ,op num (quaternion-z q))
+     :w (funcall ,op num (quaternion-w q)))))
+
+(defmacro define-quaternion-unary-op (name op)
+  `(defmethod ,name ((q quaternion))
+    (make-quaternion
+     :x (funcall ,op (quaternion-x q))
+     :y (funcall ,op (quaternion-y q))
+     :z (funcall ,op (quaternion-z q))
+     :w (funcall ,op (quaternion-w q)))))
+
+(define-quaternion-op add #'+)
+(define-quaternion-op sub #'-)
+
+(define-quaternion-num-op mul #'*)
+(define-quaternion-num-op div #'/)
+
+(define-quaternion-num-op-revord mul #'*)
+(define-quaternion-num-op-revord div #'/)
+
+(define-quaternion-unary-op absq #'abs)
+(define-quaternion-unary-op negq #'-)
+
+(defmethod add ((value real) (q quaternion))
+  (make-quaternion
+    :w (+ (quaternion-w q) value)
+    :x (quaternion-x q)
+    :y (quaternion-y q)
+    :z (quaternion-z q)))
+
+(defmethod add ((q quaternion) (value real))
+  (make-quaternion
```

Comment:
You can call add with swapped arguments here.
Or extract the common part to a non-exported macro which use in these methods.